cpp/core/shuffle/LocalPartitionWriter.cc

```diff
@@ -32,6 +33,11 @@ std::string LocalPartitionWriterBase::nextSpilledFileDir() {
 
 arrow::Status LocalPartitionWriterBase::setLocalDirs() {
   ARROW_ASSIGN_OR_RAISE(configuredDirs_, getConfiguredLocalDirs());
+  // Shuffle the configured local directories. This prevents each task from using the same directory for spilled files.
+  std::random_device rd;
+  std::default_random_engine engine(rd());
+  std::shuffle(configuredDirs_.begin(), configuredDirs_.end(), engine);
+
```

Comment:
nit: It would sound better to me if we generate a random ID for the configured directory.

Reply:
Do you mean generating a random starting index and then cycling to get the next directory, or selecting a directory randomly each time?

Reply:
The second one, selecting a directory randomly each time. Spark also use the same way to get the random directory.

Reply:
Spark uses the mod of hash of the filename as the index. Here the spilled file is only used for native writer, so I think we don't need to completely follow spark's way as the spilling mechanism are also not the same. We only need to make sure the directory is randomly chosen.

Reply:
The current idea seems fine to me. I'm just concerned about the shuffle cost for the configuredDirs, especially considering that we might have up to 10 directories in configuredDirs.

Reply:
This is a one-time op at the beginning of each shuffle write task. After shuffling, it saves the result to `configuredDirs_`, and looping to get the next directory each time.

Reply:
ok